src/query/api/v1/handler/prometheus/native/read.go

```diff
@@ -94,10 +97,24 @@ func NewPromReadHandler(
 	}
 }
 
+type errorWithID struct {
+	xhttp.ErrorResponse
+	RqID string `json:"rqID"`
+}
+
 func (h *PromReadHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+
 	result, params, respErr := h.ServeHTTPWithEngine(w, r, h.engine)
 	if respErr != nil {
-		xhttp.Error(w, respErr.Err, respErr.Code)
+		logging.ReadContextID(ctx)
+		w.WriteHeader(respErr.Code)
+		xhttp.WriteJSONResponse(w, errorWithID{
+			ErrorResponse: xhttp.ErrorResponse{
+				Error: respErr.Err.Error(),
+			},
+			RqID: logging.ReadContextID(ctx),
+		}, logging.WithContext(ctx))
```

Comment:
Do we need to log id in the error returned to the client? The failing request ID is still logged since the logger has the context id on it

Reply:
Sure, doesn't seem like it would cause issues. We've only got tracing on read right? Might want to add the same thing to `read_instantaneous`, too... Would it make sense to take this block and put it into `xhttp` as a method like `WriteJSONResponseWithRequestID`?

Reply:
Yeah good call, I shouldn't have been lazy about it here :). I'll factor it out.